sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/HashUtils.scala

```diff
@@ -0,0 +1,135 @@
+
+    override def close(): Unit = sb.clear()
+  }
+
+  /** Interface for message digest. Allows to swap out default implementation for debug implementation.
+    */
+  private[crypto] trait MessageDigestWithContext {
+    def messageDigest: MessageDigest
+    def update(a: Byte, context: Option[String]): Unit
+    def update(a: ByteBuffer, context: Option[String]): Unit
+    def update(a: Array[Byte], context: Option[String]): Unit
+    def digest(buf: Array[Byte], offset: Int, len: Int): Int
+    def addContext(context: String): Unit
+    def isContextAware: Boolean
+  }
+
+  /** Default Message Digest implementation. Ignores context and simply feeds all bytes to the delegate MessageDigest.
+    * @param messageDigest MessageDigest delegate which calculates the hash.
+    */
+  private[crypto] class DefaultMessageDigest(override val messageDigest: MessageDigest)
+      extends MessageDigestWithContext {
+    override val isContextAware: Boolean = false
+    override def update(a: Byte, context: Option[String]): Unit = messageDigest.update(a)
+    override def update(a: ByteBuffer, context: Option[String]): Unit = messageDigest.update(a)
+    override def update(a: Array[Byte], context: Option[String]): Unit = messageDigest.update(a)
+    override def digest(buf: Array[Byte], offset: Int, len: Int): Int =
+      messageDigest.digest(buf, 0, underlyingHashLength)
+    // Default message digest doesn't do anything with the context
+    override def addContext(context: String): Unit = {}
+  }
+
+  /** Message Digest for debugging with support for contextual information and output stream to observe the encoding result.
+    *
+    * @param messageDigest MessageDigest delegate which calculates the hash.
+    * @param outputStream Output stream that also receives the encoded objects along with context information.
+    *
+    * The format of the produced debug output is the following:
+    *
+    * # Context About Following Lines
+    * "encoded value in hex" - ["original value before encoding" ("additional contextual info", for instance type of the value)]
+    *
+    * e.g:
+    *   # Package Name
+    *   0000000e - [14 (int)]
+    *   7061636b6167652d6e616d652d30 - [package-name-0 (string)]
+    */
+  private[crypto] class DebugMessageDigest(
+      messageDigest: MessageDigest,
+      outputStream: ContextAwareOutputStream,
+  ) extends DefaultMessageDigest(messageDigest) {
+    override val isContextAware: Boolean = true
+
+    override def update(a: ByteBuffer, context: Option[String]): Unit = {
+      if (a.hasArray) {
+        update(a.array(), context)
+      } else {
+        // Copied from MessageDigestSpi.engineUpdate(ByteBuffer input)
+        var len = a.remaining
+        val tempArray = new Array[Byte](len)
+        while (len > 0) {
+          val chunk = Math.min(len, tempArray.length)
+          discard(a.get(tempArray, 0, chunk))
+          len -= chunk
+        }
+        update(tempArray, context)
```

Comment:
This piece of code doesn't make much sense any more. `MessageDigestSpi.engineUpdate` uses a fixed `tempArray` for all calls to `update`. Here, you create a fresh `tempArray` of the appropriate size upon each call. So the `tempArray` always has enough capacity to hold the whole array. So no need for a while loop here. Please simplify this!